File: dead_code_finder.py
import os
import re

def get_all_files(directory, extensions):
    files = []
    for root, _, filenames in os.walk(directory):
        if 'node_modules' in root or '.git' in root or '__pycache__' in root:
            continue
        for filename in filenames:
            if any(filename.endswith(ext) for ext in extensions):
                files.append(os.path.join(root, filename))
    return files

def check_unused():
    with open('unused.txt', 'w', encoding='utf-8') as out_file:
        out_file.write("--- DEAD CODE SCANNING ---\n")
        
        frontend_dir = r"c:\Uni\Projekte\f1-predictions\src"
        backend_dir = r"c:\Uni\Projekte\f1-predictions\api"
        
        ts_files = get_all_files(frontend_dir, ['.ts', '.tsx'])
        css_files = get_all_files(frontend_dir, ['.css'])
        py_files = get_all_files(backend_dir, ['.py'])
        
        all_ts_content = ""
        for f in ts_files:
            try:
                with open(f, 'r', encoding='utf-8') as file:
                    all_ts_content += file.read() + "\n"
            except Exception:
                pass
                
        out_file.write("\n[FRONTEND] Potentially unused TS/TSX files:\n")
        for f in ts_files:
            basename = os.path.basename(f)
            name_no_ext = os.path.splitext(basename)[0]
            if name_no_ext in ['main', 'App', 'vite-env.d', 'f1-presets']:
                continue
            if all_ts_content.count(name_no_ext) < 2:
                out_file.write(f"- {f}\n")

        out_file.write("\n[CSS] Potentially unused CSS files:\n")
        for f in css_files:
            basename = os.path.basename(f)
            if basename not in all_ts_content:
                out_file.write(f"- {f}\n")
                
        out_file.write("\n[CSS] Checking for unused CSS classes in F1Footer and F1Header:\n")
        for css_file in css_files:
            if 'F1Footer.css' in css_file or 'F1Header.css' in css_file:
                try:
                    with open(css_file, 'r', encoding='utf-8') as file:
                        content = file.read()
                        classes = re.findall(r'\.([a-zA-Z0-9_-]+)\s*\{', content)
                        for c in set(classes):
                            if c not in all_ts_content:
                                out_file.write(f"- Unused class '{c}' in {os.path.basename(css_file)}\n")
                except Exception:
                    pass
                    
        out_file.write("\n[BACKEND] Potentially unused Python files:\n")
        all_py_content = ""
        for f in py_files:
            try:
                with open(f, 'r', encoding='utf-8') as file:
                    all_py_content += file.read() + "\n"
            except Exception:
                pass
                
        for f in py_files:
            basename = os.path.basename(f)
            name_no_ext = os.path.splitext(basename)[0]
            if name_no_ext in ['index', 'dev_server', '__init__']:
                continue
            if all_py_content.count(name_no_ext) < 2:
                out_file.write(f"- {f}\n")

if __name__ == "__main__":
    check_unused()

File: dead_code_finder.js
const fs = require('fs');
const path = require('path');

function getAllFiles(dirPath, extensions, arrayOfFiles) {
  const files = fs.readdirSync(dirPath);
  arrayOfFiles = arrayOfFiles || [];

  files.forEach(function(file) {
    if (file === 'node_modules' || file === '.git' || file === '__pycache__') return;
    const fullPath = path.join(dirPath, file);
    if (fs.statSync(fullPath).isDirectory()) {
      arrayOfFiles = getAllFiles(fullPath, extensions, arrayOfFiles);
    } else {
      if (extensions.some(ext => file.endsWith(ext))) {
        arrayOfFiles.push(fullPath);
      }
    }
  });
  return arrayOfFiles;
}

try {
  let output = "--- DEAD CODE SCANNING ---\n";
  const frontendDir = path.join('c:', 'Uni', 'Projekte', 'f1-predictions', 'src');
  const backendDir = path.join('c:', 'Uni', 'Projekte', 'f1-predictions', 'api');

  const tsFiles = getAllFiles(frontendDir, ['.ts', '.tsx']);
  const cssFiles = getAllFiles(frontendDir, ['.css']);
  const pyFiles = getAllFiles(backendDir, ['.py']);

  let allTsContent = "";
  tsFiles.forEach(f => {
    try { allTsContent += fs.readFileSync(f, 'utf-8') + "\n"; } catch (e) {}
  });

  output += "\n[FRONTEND] Potentially unused TS/TSX files:\n";
  tsFiles.forEach(f => {
    const basename = path.basename(f);
    const nameNoExt = path.parse(basename).name;
    if (['main', 'App', 'vite-env.d', 'f1-presets'].includes(nameNoExt)) return;
    const count = (allTsContent.match(new RegExp(nameNoExt, 'g')) || []).length;
    if (count < 2) {
      output += `- ${f}\n`;
    }
  });

  output += "\n[CSS] Potentially unused CSS files:\n";
  cssFiles.forEach(f => {
    const basename = path.basename(f);
    if (!allTsContent.includes(basename)) {
      output += `- ${f}\n`;
    }
  });

  output += "\n[CSS] Checking for unused CSS classes in F1Footer and F1Header:\n";
  cssFiles.forEach(f => {
    if (f.includes('F1Footer.css') || f.includes('F1Header.css')) {
      try {
        const content = fs.readFileSync(f, 'utf-8');
        const classMatches = content.match(/\.([a-zA-Z0-9_-]+)\s*\{/g) || [];
        const classes = [...new Set(classMatches.map(c => c.replace('.', '').replace('{', '').trim()))];
        classes.forEach(c => {
          if (!allTsContent.includes(c)) {
            output += `- Unused class '${c}' in ${path.basename(f)}\n`;
          }
        });
      } catch (e) {}
    }
  });

  let allPyContent = "";
  pyFiles.forEach(f => {
    try { allPyContent += fs.readFileSync(f, 'utf-8') + "\n"; } catch (e) {}
  });

  output += "\n[BACKEND] Potentially unused Python files:\n";
  pyFiles.forEach(f => {
    const basename = path.basename(f);
    const nameNoExt = path.parse(basename).name;
    if (['index', 'dev_server', '__init__'].includes(nameNoExt)) return;
    const count = (allPyContent.match(new RegExp(nameNoExt, 'g')) || []).length;
    if (count < 2) {
      output += `- ${f}\n`;
    }
  });

  fs.writeFileSync('unused.txt', output, 'utf-8');
  console.log("Done checking dead code");
} catch (e) {
  console.error(e);
}
